### DISCUSSION TITLE
time-correlated constraints

### DISCUSSION THREAD

**Discussion by P11GH** [do-mpc/do-mpc#123] [Q&A] [answered]
**Is your feature request related to a problem? Please describe.**
I need a constraint such that u(1)+u(2)+...+u(n)<= C, where u is the input variable, 1, 2,...,n are time steps in MPC prediction horizon, and C is the constant.

**Describe the solution you'd like**
I am not sure whether do-mpc has this feature on time-correlated constraints. If yes, could you add it to the document. If no, could you add it to the package?

**Additional context**
nope

**Comment by 4flixt** [comment, marked as answer]
Hi P11GH,

apologies for the late reply. There is a straight forward solution for this problem: You can just add a new artificial state with the following dynamics:

```python
u = model.set_variable('_u', 'u', 3) # your inputs
u_sum = model.set_variable('_x','u_sum', u.shape[0]) # a state that represents the sum of the inputs

# Discrete model: just add the current input to the sum of inputs
model.set_rhs('u_sum', u+u_sum)

# Continous model: Integrate (constant) u over dt and divide by dt to add current u to u_sum.
model.set_rhs('u_sum', u/dt)
```

Now you want to ensure that `` u(1)+u(2)+...+u(n)<= C``. Your new variable ``u_sum`` exists at every time-step of the horizon and represent the sum of inputs until this timestep. If you just want to penalize the sum at the final time-step you could use the [terminal constraints](https://www.do-mpc.com/en/latest/api/do_mpc.controller.MPC.terminal_bounds.html#do_mpc.controller.MPC.terminal_bounds):

```python
mpc.terminal_bounds['upper', 'u_sum'] = C
```

This should do the trick nicely. 

Hope this helps.

Best 
4flixt

**Reply by P11GH** [reply]
It's a smart way. Thank you very much.
